sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -0,0 +1,451 @@
+    }
+  }
+
+  /**
+   * Commit all the updates made as a version to DFS. The steps it needs to do to commits are:
+   * - Write all the updates to the native RocksDB
+   * - Flush all changes to disk
+   * - Create a RocksDB checkpoint in a new local dir
+   * - Sync the checkpoint dir files to DFS
+   */
+  def commit(): Long = {
+    val newVersion = loadedVersion + 1
+    val checkpointDir = createTempDir("checkpoint")
+    try {
+      // Make sure the directory does not exist. Native RocksDB fails if the directory to
+      // checkpoint exists.
+      Utils.deleteRecursively(checkpointDir)
+
+      logInfo(s"Writing updates for $newVersion")
+      val writeTimeMs = timeTakenMs { db.write(writeOptions, writeBatch) }
+
+      logInfo(s"Flushing updates for $newVersion")
+      val flushTimeMs = timeTakenMs { db.flush(flushOptions) }
+
+      val compactTimeMs = if (conf.compactOnCommit) {
+        logInfo("Compacting")
+        timeTakenMs { db.compactRange() }
+      } else 0
+      logInfo("Pausing background work")
+
+      val pauseTimeMs = timeTakenMs {
+        db.pauseBackgroundWork() // To avoid files being changed while committing
+      }
+
+      logInfo(s"Creating checkpoint for $newVersion in $checkpointDir")
+      val checkpointTimeMs = timeTakenMs {
+        val cp = Checkpoint.create(db)
+        cp.createCheckpoint(checkpointDir.toString)
+      }
+
+      logInfo(s"Syncing checkpoint for $newVersion to DFS")
+      val fileSyncTimeMs = timeTakenMs {
+        fileManager.saveCheckpointToDfs(checkpointDir, newVersion, numUncommittedKeys)
+      }
+      numCommittedKeys = numUncommittedKeys
+      loadedVersion = newVersion
+      commitLatencyMs ++= Map(
+        "writeBatch" -> writeTimeMs,
+        "flush" -> flushTimeMs,
+        "compact" -> compactTimeMs,
+        "pauseBg" -> pauseTimeMs,
+        "checkpoint" -> checkpointTimeMs,
+        "fileSync" -> fileSyncTimeMs
+      )
+      loadedVersion
+    } catch {
+      case t: Throwable =>
+        loadedVersion = -1  // invalidate loaded version
+        throw t
+    } finally {
+      db.continueBackgroundWork()
+      silentDeleteRecursively(checkpointDir, s"committing $newVersion")
+      release()
+    }
+  }
+
+  /**
+   * Drop uncommitted changes, and roll back to previous version.
+   */
+  def rollback(): Unit = {
+    writeBatch.clear()
+    numUncommittedKeys = numCommittedKeys
+    release()
+    logInfo(s"Rolled back to $loadedVersion")
+  }
+
+  /** Release all resources */
+  def close(): Unit = {
+    try {
+      closeDB()
+
+      // Release all resources related to native RockDB objects
+      writeBatch.clear()
+      writeBatch.close()
+      readOptions.close()
+      writeOptions.close()
+      flushOptions.close()
+      dbOptions.close()
+      dbLogger.close()
+      silentDeleteRecursively(localRootDir, "stopping")
+    } catch {
+      case e: Exception =>
+        logWarning("Error closing RocksDB", e)
```

Comment:
If we want to do logging only at all `Exception`, `Utils.tryLogNonFatalError` might be better.

Reply:
Thanks for the advice. But since NonFatal doesn't include `InterruptedException` which is also an Exception, maybe we'd better not change to use `Utils.tryLogNonFatalError` here.